axum-extra/src/multipart_builder.rs

```diff
@@ -0,0 +1,337 @@
+//! Generate forms to use in responses. You're probably looking for [MultipartForm].
+
+use axum::response::{IntoResponse, Response};
+use http::{header, HeaderMap};
+use std::time::{SystemTime, UNIX_EPOCH};
+
+/// The `Content-Transfer-Encoding` setting for a part.
+#[derive(Debug)]
+pub enum TransferEncoding {
+    /// If not specified, encoding defaults to UTF-8
+    Default,
+    /// If transferring raw binary data that is not guaranteed to be valid UTF-8.
+    Binary,
+}
+
+/// Create multipart forms to be used in API responses.
+/// This struct implements [IntoResponse], and so it can be returned from a handler like normal.
+#[derive(Debug)]
+pub struct MultipartForm {
+    parts: Vec<Part>,
+}
+
+impl MultipartForm {
+    /// Construct a new empty multipart form with no parts.
+    pub fn new() -> Self {
+        MultipartForm { parts: Vec::new() }
+    }
+
+    /// Initialize a new multipart form with the provided vector of parts.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use axum_extra::multipart_builder::{MultipartForm, Part};
+    ///
+    /// let parts: Vec<Part> = vec![Part::text("foo", "abc"), Part::text("bar", "def")];
+    /// let form = MultipartForm::with_parts(parts);
+    /// ```
+    pub fn with_parts(parts: Vec<Part>) -> Self {
+        MultipartForm { parts }
+    }
+
+    /// Add a new [Part] to the form
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use axum_extra::multipart_builder::{MultipartForm, Part};
+    ///
+    /// let mut form = MultipartForm::new();
+    /// form
+    ///     .part(Part::text("foo", "abc"))
+    ///     .part(Part::text("other_field_name", "def"))
+    ///     .part(Part::file("file", "file.txt", vec![0x68, 0x68, 0x20, 0x6d, 0x6f, 0x6d]));
+    /// ```
+    pub fn part(&mut self, part: Part) -> &mut Self {
+        self.parts.push(part);
+        self
+    }
+}
+
+impl IntoResponse for MultipartForm {
+    fn into_response(self) -> Response {
+        // see RFC2388 for details
+        let boundary = generate_boundary();
+        let mut headers = HeaderMap::new();
+        headers.insert(
+            header::CONTENT_TYPE,
+            format!("multipart/form-data; boundary={}", boundary)
+                .parse()
+                .unwrap(),
```

Comment:
I would caution against unwraps. I think this should not happen, but either write a comment about why it's totally ok or just bite the bullet and return an explicit 500 here on error.